feat: add accessor protocol support and refactor `stats/base/nanvariancech`

rahulptl165 opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: na
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Resolves #5674 .

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds `accessors.js` files.
-  Modified `lib/nanvariancech`.
-  Modified `lib/ndarray.js`
-  Modified readme, examples, benchmark and docs file.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #5674 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

rahulptl165 committed:
feat: add accessor protocol support and refactor stats/base/nanvarinacech

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/README.md:
```diff
@@ -134,41 +131,35 @@ Note that indexing is relative to the first index. To introduce an offset, use [
-var N = floor( x0.length / 2 );
-
-var v = nanvariancech( N, 1, x1, 2 );
+var v = nanvariancech( 4, 1, x1, 2 );
```
```suggestion
var v = nanvariancech( 5, 1, x1, 2 );
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/README.md:
```diff
@@ -114,17 +114,14 @@ The function has the following parameters:
-var N = floor( x.length / 2 );
 
-var v = nanvariancech( N, 1, x, 2 );
+var v = nanvariancech( 4, 1, x, 2 );
```
```suggestion
var v = nanvariancech( 5, 1, x, 2 );
```
we should be traversing over the `NaN` value so that the example can demonstrate how the algorithm deals with it

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/docs/repl.txt:
```diff
@@ -1,10 +1,10 @@
 
-    The `N` and `stride` parameters determine which elements in `x` are accessed
-    at runtime.
+    The `N` and stride parameters determine which elements in the strided arrays
```
```suggestion
    The `N` and stride parameters determine which elements in the strided array
```
there's only 1 array here

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/docs/repl.txt:
```diff
@@ -46,31 +46,28 @@
     ~4.3333
 
     // Using `N` and `stride` parameters:
     > x = [ -2.0, 1.0, 1.0, -5.0, 2.0, -1.0 ];
```
Add a `NaN` element in the array and traverse over it to show that we're ignoring it

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/docs/repl.txt:
```diff
@@ -46,31 +46,28 @@
     ~4.3333
 
     // Using view offsets:
     > var x0 = new {{alias:@stdlib/array/float64}}( [ 1.0, -2.0, 3.0, 2.0, 5.0, -1.0 ] );
```
similar comment regarding `NaN` values

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/docs/repl.txt:
```diff
@@ -46,31 +46,28 @@
     While typed array views mandate a view offset based on the underlying
-    buffer, the `offset` parameter supports indexing semantics based on a
-    starting index.
+    buffer, the offset parameters support indexing semantics based on a
```
```suggestion
    buffer, the offset parameter supports indexing semantics based on a
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/docs/repl.txt:
```diff
@@ -46,31 +46,28 @@
-    buffer, the `offset` parameter supports indexing semantics based on a
-    starting index.
+    buffer, the offset parameters support indexing semantics based on a
+    starting indices.
```
```suggestion
    starting index.
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/docs/repl.txt:
```diff
@@ -108,13 +105,12 @@
     ~4.3333
 
     // Using offset parameter:
     > var x = [ 1.0, -2.0, 3.0, 2.0, 5.0, -1.0 ];
```
similar comment regarding `NaN` values

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/test/test.nanvariancech.js:
```diff
@@ -230,15 +352,36 @@ tape( 'the function supports a `stride` parameter', function test( t ) {
+		NaN
+	];
+
+	v = nanvariancech( 4, 1, toAccessorArray( x ), 2 );
```
```suggestion
	v = nanvariancech( 5, 1, toAccessorArray( x ), 2 );
```
notice in the comments that the indices which should be traversed are mentioned, kindly follow that

aayush0325 replied:
same for all other instances.

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/test/test.nanvariancech.js:
```diff
@@ -230,15 +352,36 @@ tape( 'the function supports a `stride` parameter', function test( t ) {
 
-	N = floor( x.length / 2 );
-	v = nanvariancech( N, 1, x, 2 );
+	v = nanvariancech( 4, 1, x, 2 );
```
```suggestion
	v = nanvariancech( 5, 1, x, 2 );
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/test/test.nanvariancech.js:
```diff
@@ -313,9 +485,8 @@ tape( 'the function supports view offsets', function test( t ) {
-	N = floor(x1.length / 2);
 
-	v = nanvariancech( N, 1, x1, 2 );
+	v = nanvariancech( 4, 1, x1, 2 );
```
```suggestion
	v = nanvariancech( 5, 1, x1, 2 );
```
kindly follow the commented indices, same comment for all other instances

aayush0325 commented:
kindly resolve the merge conflicts in readme as well

rahulptl165 committed:
refactor: add requested changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 commented:
conflicts in README haven't been resolved

aayush0325 commented:
kindly match the file structure that we're following, your `lib` directory should only contain an `index.js`, `main.js`, `ndarray.js` and `accessors.js`. see #6248 and [nanrange/lib](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/stats/base/nanrange/lib) for example

rahulptl165 committed:
refactor: refactor the lib folder's files to align with existing style

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

rahulptl165 commented:
Hi @aayush0325 , I have addressed requested changes. let me know if you want further modification.

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/README.md:
```diff
@@ -183,7 +174,7 @@ var v = nanvariancech.ndarray( N, 1, x, 2, 1 );
 -   If `N <= 0`, both functions return `NaN`.
```
no note for support of accessor arrays here

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/lib/main.js:
```diff
@@ -20,14 +20,41 @@
+* // returns ~4.3333
+*/
+function nanvariancech( N, correction, x, strideX ) {
+	return ndarray( N, correction, x, strideX, stride2offset( N, strideX) );
```
```suggestion
	return ndarray( N, correction, x, strideX, stride2offset( N, strideX ) );
```

rahulptl165 committed:
docs: add note about accessor array support

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

rahulptl165 commented:
Hi @aayush0325 , addressed requested changes.

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into feat/accessor-nanvariancech

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/README.md:
```diff
@@ -98,7 +98,7 @@ The use of the term `n-1` is commonly referred to as Bessel's correction. Note,
-#### nanvariancech( N, correction, x, stride )
+#### nanvariancech( N, correction, x, strideX )
 
 Computes the [variance][variance] of a strided array `x` ignoring `NaN` values and using a one-pass trial mean algorithm.
```
```suggestion
Computes the [variance][variance] of a strided array ignoring `NaN` values and using a one-pass trial mean algorithm.
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/README.md:
```diff
@@ -114,61 +114,52 @@ The function has the following parameters:
 var x = [ 1.0, -2.0, NaN, 2.0 ];
 
-var v = nanvariancech.ndarray( x.length, 1, x, 1, 0 );
+var v = nanvariancech.ndarray( 4, 1, x, 1, 0 );
```
```suggestion
var v = nanvariancech.ndarray( x.length, 1, x, 1, 0 );
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/docs/repl.txt:
```diff
@@ -46,30 +46,27 @@
     // Standard Usage:
     > var x = [ 1.0, -2.0, NaN, 2.0 ];
-    > {{alias}}( x.length, 1, x, 1 )
+    > {{alias}}( 4, 1, x, 1 )
```
```suggestion
    > {{alias}}( x.length, 1, x, 1 )
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/docs/repl.txt:
```diff
@@ -46,30 +46,27 @@
+    > {{alias}}( 4, 1, x, 1 )
     ~4.3333
 
     // Using `N` and `stride` parameters:
```
```suggestion
    // Using `N` and stride parameters:
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/docs/repl.txt:
```diff
@@ -108,13 +105,12 @@
     // Standard Usage:
     > var x = [ 1.0, -2.0, NaN, 2.0 ];
-    > {{alias}}.ndarray( x.length, 1, x, 1, 0 )
+    > {{alias}}.ndarray( 4, 1, x, 1, 0 )
```
```suggestion
    > {{alias}}.ndarray( x.length, 1, x, 1, 0 )
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/lib/accessors.js:
```diff
@@ -36,17 +36,25 @@
 */
-function nanvariancech( N, correction, x, stride ) {
+function nanvariancech( N, correction, x, strideX, offsetX ) {
+	var xget;
```
xget should be renamed as get

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/lib/index.js:
```diff
@@ -28,23 +28,28 @@
 * var x = [ 1.0, -2.0, NaN, 2.0 ];
 *
-* var v = nanvariancech( x.length, 1, x, 1 );
+* var v = nanvariancech( 4, 1, x, 1 );
```
```suggestion
* var v = nanvariancech( x.length, 1, x, 1 );
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/lib/main.js:
```diff
@@ -20,14 +20,41 @@
+* @example
+* var x = [ 1.0, -2.0, NaN, 2.0 ];
+*
+* var v = nanvariancech( 4, 1, x, 1 );
```
```suggestion
* var v = nanvariancech( x.length, 1, x, 1 );
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/test/test.main.js:
```diff
@@ -148,6 +148,114 @@ tape( 'the function calculates the sample variance of a strided array (ignoring
 	t.end();
 });
 
+tape( 'the function calculates the population variance of a strided array (accessors) (ignoring `NaN` values)', function test( t ) {
```
```suggestion
tape( 'the function calculates the population variance of a strided array (ignoring `NaN` values) (accessors)', function test( t ) {
```

gururaj1512 commented on lib/node_modules/@stdlib/stats/base/nanvariancech/test/test.main.js:
```diff
@@ -163,6 +271,21 @@ tape( 'if provided an `N` parameter less than or equal to `0`, the function retu
 	t.end();
 });
 
+tape( 'if provided an `N` parameter less than or equal to `0`, the function returns `NaN`', function test( t ) {
```
```suggestion
tape( 'if provided an `N` parameter less than or equal to `0`, the function returns `NaN` (accessors)', function test( t ) {
```

gururaj1512 committed:
test: update variance tests to use accessors

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
fix: update nanvariancech example

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
fix: update examples in README and REPL for nanvariancech

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 reviewed (dismissed):
LGTM. onto @kgryte

gururaj1512 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into feat/accessor-nanvariancech

kgryte commented on lib/node_modules/@stdlib/stats/base/nanvariancech/README.md:
```diff
@@ -158,17 +152,14 @@ var v = nanvariancech.ndarray( x.length, 1, x, 1, 0 );
+-   **offsetX**: starting index for `x`.
 
-While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on a starting index. For example, to calculate the [variance][variance] for every other value in `x` starting from the second value
+While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on a starting index. For example, to calculate the [variance][variance] for every other element in the strided array starting from the second element
```
```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameter supports indexing semantics based on a starting index. For example, to calculate the [variance][variance] for every other element in the strided array starting from the second element
```

kgryte committed:
docs: remove backticks

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/nanvariancech/docs/repl.txt:
```diff
@@ -112,9 +109,8 @@
-    > var x = [ 1.0, -2.0, 3.0, 2.0, 5.0, -1.0 ];
-    > var N = {{alias:@stdlib/math/base/special/floor}}( x.length / 2 );
-    > {{alias}}.ndarray( N, 1, x, 2, 1 )
+    > var x = [ 1.0, -2.0, 3.0, 2.0, 5.0, -1.0, NaN, NaN ];
```
```suggestion
    > x = [ 1.0, -2.0, 3.0, 2.0, 5.0, -1.0, NaN, NaN ];
```

kgryte committed:
docs: avoid duplicate declaration

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/nanvariancech/lib/accessors.js:
```diff
@@ -36,17 +36,25 @@
 *
```
```suggestion
*
* @private
```

kgryte committed:
docs: add missing private annotation

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
